tests/cuda_ndarrays/conftest.py

```diff
@@ -0,0 +1,132 @@
+
+        subprocess.run(comp_cmd, check= 'TRUE')
+        if sys.platform.startswith("win"):
+            test_exe += ".exe"
+        test_output = subprocess.check_output("./" + test_exe)
+
+        # Clean up the unit test output and remove non test data lines.
+        lines = test_output.decode().split("\n")
+        lines = [line.strip() for line in lines]
+        lines = [line for line in lines if line.startswith("[")]
+
+        # Extract the test metadata from the unit test output.
+        test_results = []
+        for line in lines:
+            token, data = line.split(" ", 1)
+            token = token[1:-1]
+            if token in ("PASS", "FAIL"):
+                file_name, line_number, function_name = data.split(":")
+                test_results.append({"condition": token,
+                                     "file_name": file_name,
+                                     "function_name": function_name,
+                                     "line_number": int(line_number),
+                                     "INFO" : "no data found",
+                                     "DSCR" : ""
+                                     })
+            elif token in ("INFO", "DSCR"):
+                test_results[-1][token] = data
+        for test_result in test_results:
+            if NEEDS_FROM_PARENT:
+                yield CTestItem.from_parent(name = test_result["function_name"],
+                        parent = self, test_result = test_result)
+            else:
+                yield CTestItem(name = test_result["function_name"], parent = self,
+                        test_result = test_result)
+
+
+class CTestItem(pytest.Item):
+    """
+    Pytest.Item subclass to handle each test result item. There may be
+    more than one test result from a test function.
+
+    """
+
+    def __init__(self, *, test_result, **kwargs):
+        """Overridden constructor to pass test results dict."""
+        super().__init__(**kwargs)
+        self.test_result = test_result
+
+    @classmethod
+    def from_parent(cls, *, test_result, **kwargs):
+        return super().from_parent(test_result=test_result, **kwargs)
+
+    def runtest(self):
+        """The test has already been run. We just evaluate the result."""
+        if self.test_result["condition"] == "FAIL":
+            raise CTestException(self, self.name)
+
+    def reportinfo(self):
+        """"Called to display header information about the test case."""
+        return self.fspath, self.test_result["line_number"], self.name
+
+    def repr_failure(self, exception):
+        """
+        Called when runtest() raises an exception. The method is used
+        to format the output of the failed test result.
+
+        """
+        if isinstance(exception.value, CTestException):
+            return (f"Test failed : {self.test_result['file_name']}:{self.test_result['line_number']} {self.test_result['function_name']} < {self.test_result['DSCR']} >\n INFO : {self.test_result['INFO']}")

```

Comment:
Have you come across other Exception types which should be ignored? Which ones? Would it not be safer to handle these types separately?

Reply:
this was left from my testing to raise different error messages.
there should not be any other exceptions possible because we are the one raising them.

Reply:
If there is a typo in the code which raises the error then a variety of errors (e.g. `NameError`) could be raised. This is very similar to [this error](https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/broad-exception-caught.html). In this case as I understand it pytest will call this function and the error will be ignored. That is why it is dangerous to ignore all other errors. I suggest:

```suggestion
        if isinstance(exception.value, CTestException):
            return (f"Test failed : {self.test_result['file_name']}:{self.test_result['line_number']} {self.test_result['function_name']} < {self.test_result['DSCR']} >\n INFO : {self.test_result['INFO']}")
        else:
             return super().repr_failure(exception)
```

Reply:
Is there no `super().repr_failure` function?

Reply:
I did make the requested changes, I just wanted to run some tests before committing them. `super.repr_failure()` is available and I have modified the code here d65ca5218b54097980aab6c7f8b4efa35193b531.

Reply:
Perfect thanks very much for fixing all this. I didn't realise the other conftest file was such a mess. We mustn't have reviewed it properly :(

Reply:
No problem at all :sweat_smile: